Add ORT inference (#113)
huggingface/optimum: docs/source/pipelines.mdx

<!--Copyright 2022 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

> # Optimum Pipelines for inference

Reviewer: ```suggestion
# Optimum pipelines for inference
```

The [`optimum_pipeline`] makes it simple to use models from the [Model Hub](https://huggingface.co/models) for accelerated inference on a variety of tasks such as text classification.

> Even if you don't have experience with a specific modality or understand the code powering the models, you can still use them with the [`optimum_pipeline`]! This tutorial will teach you to:

Reviewer: It looks like a sentence is missing after "This tutorial ..." but I realise this PR is WIP so ignore my comment if necessary :)

> <Tip>

Reviewer: Should this tip be moved below, i.e. just after the example in `Optimum pipeline usage`?

You can also use the `transformers.pipeline` and provide your `OptimumModel`.

</Tip>

Currenlty supported tasks are:

**Onnxruntime**

* `feature-extraction`
* `text-classification`
* `token-classification`
* `question-answering`
* `zero-shot-classification`

> ## Optimum Pipeline usage

Reviewer: If we're following the conventions from the `transformers` docs, then I think the convention is to lowercase everything after the first word:
```suggestion
## Optimum pipeline usage
```

While each task has an associated [`optimum_pipeline`], it is simpler to use the general [`pipeline`] abstraction which contains all the specific task pipelines. 
The [`optimum_pipeline`] automatically loads a default model and tokenizer capable of inference for your task. 

1. Start by creating a [`optimum_pipeline`] and specify an inference task:

```py
>>> from optimum import optimum_pipeline

>>> classifier = optimum_pipeline(task="text-classification", accelerator="onnx")

```

2. Pass your input text to the [`optimum_pipeline`]:

```python
>>> classifier("I like you. I love you.")
```

### Use a vanilla transformers model and convert

The [`optimum_pipeline`] accepts supported model from the [Model Hub](https://huggingface.co/models). 
There are tags on the Model Hub that allow you to filter for a model you'd like to use for your task. 

> Once you've picked an appropriate model, load it with the `.from_transformers` method for corresponding `OnnxFor*` 

Reviewer: Suggestion to use parentheses to indicate a method instead of `.` (this is apparently the norm in the publishing world):

```suggestion
Once you've picked an appropriate model, load it with the `from_transformers()` method for corresponding `OnnxFor*` 
```

>>> model = ORTModelForQuestionAnswering.from_pretrained("optimum/roberta-base-squad2")

>>> onnx_qa = optimum_pipeline("question-answering", model=model, tokenizer=tokenizer)
>>> question = "Whats my name?"
>>> context = "My Name is Philipp and I live in Nuremberg."

>>> pred = onnx_qa(question=question, context=context)
```

### Optimizing and Quantizing in Pipelines

The [`optimum_pipeline`] can not only run inference it also provides arguments to quantize and optimize your model on the fly.
Once you've picked an appropriate model, load it with the `.from_pretrained` method for corresponding `ORTModelFor*` 
and [`AutoTokenizer'] class. For example, load the [`ORTModelForQuestionAnswering`] class for a question answering modeling task and provide 
the `do_optimization=True` and/or `do_quantization=True` arguments:

```py
>>> from transformers import AutoTokenizer
>>> from optimum.onnxruntime import ORTModelForQuestionAnswering
>>> from optimum import optimum_pipeline

>>> tokenizer = AutoTokenizer.from_pretrained("optimum/roberta-base-squad2")
>>> model = ORTModelForQuestionAnswering.from_pretrained("optimum/roberta-base-squad2")

>>> model.optimize()

> >>> model.quantize()

Reviewer 2: I think this is redundant with `do_optimization=True` and `do_quantization=True` at lines 120 and 121. I guess we want to delete these lines and keep lines 120-121 since this is an example of quantization/optimization inside the pipeline.

<!--Copyright 2022 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# Optimum pipelines for inference

The [`optimum_pipeline`] makes it simple to use models from the [Model Hub](https://huggingface.co/models) for accelerated inference on a variety of tasks such as text classification.
Even if you don't have experience with a specific modality or understand the code powering the models, you can still use them with the [`optimum_pipeline`]! This tutorial will teach you to:

<Tip>

> You can also use the `transformers.pipeline` and provide your `OptimumModel`.

Reviewer: ```suggestion
You can also use the `pipeline()` function from Transformers and provide your `OptimumModel`.
```

</Tip>

Currenlty supported tasks are:

> **Onnxruntime**

Reviewer: turbo nit:

```suggestion
**Onnx Runtime**
```

* `feature-extraction`
* `text-classification`
* `token-classification`
* `question-answering`
* `zero-shot-classification`
* `text-generation`

## Optimum pipeline usage

> While each task has an associated [`optimum_pipeline`], it is simpler to use the general [`pipeline`] abstraction which contains all the specific task pipelines. 

Reviewer: I'm not sure I understand what you mean here with the difference between `optimum_pipeline` and `pipeline`. You say one should use the `pipeline` abstraction but the example following this uses `optimum_pipeline`. Perhaps we can clarify this?

Author: Copied it from here and forgot to adjust: https://huggingface.co/docs/transformers/pipeline_tutorial#pipeline-usage

The [`optimum_pipeline`] automatically loads a default model and tokenizer capable of inference for your task. 

1. Start by creating a [`optimum_pipeline`] and specify an inference task:

```py
>>> from optimum import optimum_pipeline

>>> classifier = optimum_pipeline(task="text-classification", accelerator="ort")

```

2. Pass your input text to the [`optimum_pipeline`]:

```python

> >>> classifier("I like you. I love you.")

Reviewer: You can show the outputs on a line below by not including the `>>>` delimiters if you want

```

> ### Use a vanilla transformers model and convert

Reviewer: Not sure if you meant "convert to generic IR" so feel free to drop the ONNX part
```suggestion
### Use a vanilla Transformers model and converting to ONNX
```

> The [`optimum_pipeline`] accepts supported model from the [Model Hub](https://huggingface.co/models). 

Reviewer: nit:
```suggestion
The [`optimum_pipeline`] accepts any supported model from the [Model Hub](https://huggingface.co/models). 
```

There are tags on the Model Hub that allow you to filter for a model you'd like to use for your task. 

> Once you've picked an appropriate model, load it with the `from_pretrained("{model_id}",from_transformers=True)` method for corresponding `ORTModelFor*` 

Reviewer: ```suggestion
Once you've picked an appropriate model, load it with the `from_pretrained("{model_id}",from_transformers=True)` method associated with the corresponding `ORTModelFor*` 
```

> and [`AutoTokenizer'] class. For example, load the [`ORTModelForQuestionAnswering`] class for a question answering modeling task:

Reviewer: ```suggestion
and [`AutoTokenizer'] class. For example, here's how you can load the [`ORTModelForQuestionAnswering`] class for question answering:
```

```py
>>> from transformers import AutoTokenizer
>>> from optimum.onnxruntime import ORTModelForQuestionAnswering
>>> from optimum import optimum_pipeline

>>> tokenizer = AutoTokenizer.from_pretrained("deepset/roberta-base-squad2")

> >>> model = ORTModelForQuestionAnswering.from_pretrained("deepset/roberta-base-squad2",from_transformers=True)

Reviewer: Maybe add a comment here that we load and convert here? It took me a second to spot the difference with the section below on using optimised models

>>> onnx_qa = optimum_pipeline("question-answering", model=model, tokenizer=tokenizer)

> >>> question = "Whats my name?"

Reviewer: turbo nit:
```suggestion
>>> question = "What's my name?"
```

> >>> context = "My Name is Philipp and I live in Nuremberg."

Reviewer: ```suggestion
>>> context = "My name is Philipp and I live in Nuremberg."
```

>>> pred = onnx_qa(question=question, context=context)
```

> ### Use optimum model

Reviewer: Suggestion to use same tense as in the rest of the docs
```suggestion
### Using Optimum models
```

> The [`optimum_pipeline`] is tightly integrated with [Model Hub](https://huggingface.co/models) and can load optimized models directly, e.g. Onnxruntime. 

Reviewer: ```suggestion
The [`optimum_pipeline`] is tightly integrated with [Model Hub](https://huggingface.co/models) and can load optimized models directly, e.g. those created with OnnxRuntime. 
```

There are tags on the Model Hub that allow you to filter for a model you'd like to use for your task. 

> Once you've picked an appropriate model, load it with the `.from_pretrained` method for corresponding `ORTModelFor*` 

Reviewer: ```suggestion
Once you've picked an appropriate model, load it with the `from_pretrained()` method associated with the corresponding `ORTModelFor*` 
```

> and [`AutoTokenizer'] class. For example, load the [`ORTModelForQuestionAnswering`] class for a question answering modeling task:

Reviewer: ```suggestion
and [`AutoTokenizer'] class. For example, here's how you can load an optimized model for question answering:
```

```py
>>> from transformers import AutoTokenizer
>>> from optimum.onnxruntime import ORTModelForQuestionAnswering
>>> from optimum import optimum_pipeline

>>> tokenizer = AutoTokenizer.from_pretrained("optimum/roberta-base-squad2")

> >>> model = ORTModelForQuestionAnswering.from_pretrained("optimum/roberta-base-squad2")

Reviewer: Maybe add a comment that this is an optimized model?

>>> onnx_qa = optimum_pipeline("question-answering", model=model, tokenizer=tokenizer)

> >>> question = "Whats my name?"

Reviewer: ```suggestion
>>> question = "What's my name?"
```

> >>> context = "My Name is Philipp and I live in Nuremberg."

Reviewer: ```suggestion
>>> context = "My name is Philipp and I live in Nuremberg."
```

>>> optimizer = ORTOptimizer.from_pretrained(model_id, feature="sequence-classification")
>>> optimization_config = OptimizationConfig(optimization_level=2)

# apply the optimization configuration to the model
>>> optimizer.export(
    onnx_model_path=save_path / "model.onnx",
    onnx_optimized_model_output_path=save_path / "model-optimized.onnx",
    optimization_config=optimization_config,
)
>>> optimizer.model.config.save_pretrained(save_path) # saves config.json 

# load optimized model from local path or repository
>>> model = ORTModelForSequenceClassification.from_pretrained(save_path,file_name="model-optimized.onnx")

# create transformers pipeline
>>> onnx_clx = optimum_pipeline("text-classification", model=model, tokenizer=tokenizer)
>>> text = "I like the new ORT pipeline"
>>> pred = onnx_clx(text)
>>> print(pred)

# save model & push model to the hub
>>> tokenizer.save_pretrained("new_path_for_directory")
>>> model.save_pretrained("new_path_for_directory")
>>> model.push_to_hub("new_path_for_directory",
                  repository_id="my-onnx-repo",
                  use_auth_token=True)
```

## Transformers pipeline usage

> The [`optimum_pipeline`] is just a light wrapper around the `transformers.pipeline` to enable checks for supported tasks and additional features

Reviewer: ```suggestion
The [`optimum_pipeline`] is just a light wrapper around the `transformers.pipeline` function to enable checks for supported tasks and additional features
```

, like quantization and optimization. This being said you can use the `transformers.pipeline` and just replace your `AutoFor*` with the optimum 
 `ORTModelFor*` class. 

```diff
from transformers import AutoTokenizer, pipeline
-from transformers import AutoModelForQuestionAnswering
+from optimum.onnxruntime import ORTModelForQuestionAnswering

-model = AutoModelForQuestionAnswering.from_pretrained("deepset/roberta-base-squad2")
+model = ORTModelForQuestionAnswering.from_transformers("optimum/roberta-base-squad2")
tokenizer = AutoTokenizer.from_pretrained("deepset/roberta-base-squad2")

onnx_qa = pipeline("question-answering",model=model,tokenizer=tokenizer)

> question = "Whats my name?"

Reviewer: ```suggestion
question = "What's my name?"
```

> context = "My Name is Philipp and I live in Nuremberg."

Reviewer: ```suggestion
context = "My name is Philipp and I live in Nuremberg."
```

Even if you don't have experience with a specific modality or understand the code powering the models, you can still use them with the [`optimum_pipeline`]! This tutorial will teach you to:

<Tip>

You can also use the `pipeline()` function from Transformers and provide your `OptimumModel`.

</Tip>

Currenlty supported tasks are:

**Onnx Runtime**

* `feature-extraction`
* `text-classification`
* `token-classification`
* `question-answering`
* `zero-shot-classification`
* `text-generation`

## Optimum pipeline usage

While each task has an associated [~`optimum_pipeline`], which it is simpler to use the general [~`optimum_pipeline`] abstraction which contains all the specific task pipelines. 
The [~`optimum_pipeline`] automatically loads a default model and tokenizer capable of inference for your task. 

1. Start by creating a [~`optimum_pipeline`] and specify an inference task:

```python
>>> from optimum import optimum_pipeline

>>> classifier = optimum_pipeline(task="text-classification", accelerator="ort")

```

2. Pass your input text to the [~`optimum_pipeline`]:

```python
>>> classifier("I like you. I love you.")
[{'label': 'POSITIVE', 'score': 0.9998838901519775}]
```

> _Note: The default models used in the [~`optimum_pipeline`] are not optimized or quantized, there won't be an performance improvement comapred to there pytorch counter parts._

Reviewer 3: Typo -> compared
